xdr/json.go

```diff
@@ -3,15 +3,62 @@ package xdr
 import (
 	"encoding/json"
 	"fmt"
+	"regexp"
+	"strconv"
 	"time"
+
+	"github.com/stellar/go/support/errors"
 )
 
+// ISO8601Time is a timestamp which supports parsing dates which have a year outside the 0000..9999 range
+type ISO8601Time struct {
+	time.Time
+}
+
+// reISO8601 is the regular expression used to parse date strings in the
+// ISO 8601 extended format, with or without an expanded year representation.
+var reISO8601 = regexp.MustCompile(`^([-+]?\d{4,})-(\d{2})-(\d{2})`)
+
+// MarshalJSON serializes the timestamp to a string
+func (t ISO8601Time) MarshalJSON() ([]byte, error) {
+	return []byte(fmt.Sprintf(`"%s"`, t.Format(time.RFC3339))), nil
+}
+
+// UnmarshalJSON parses a JSON string into a ISO8601Time instance.
+func (t *ISO8601Time) UnmarshalJSON(b []byte) error {
+	if string(b) == "null" {
+		return nil
+	}
+	if b[0] != '"' || b[len(b)-1] != '"' {
+		return fmt.Errorf("%s does not begin and end with double quotes", b)
+	}
+	trimmed := string(b[1 : len(b)-1])
+	m := reISO8601.FindStringSubmatch(trimmed)
+
+	if len(m) != 4 {
+		return fmt.Errorf("UnmarshalJSON: cannot parse %s", trimmed)
+	}
+	// No need to check for errors since the regexp guarantees the matches
+	// are valid integers
+	year, _ := strconv.Atoi(m[1])
+	month, _ := strconv.Atoi(m[2])
+	day, _ := strconv.Atoi(m[3])
+
+	ts, err := time.Parse(time.RFC3339, "2006-01-02"+trimmed[len(m[0]):])
+	if err != nil {
+		return errors.Wrap(err, "Could not extract time")
+	}
+
+	t.Time = time.Date(year, time.Month(month), day, ts.Hour(), ts.Minute(), ts.Second(), ts.Nanosecond(), ts.Location())
+	return nil
```

Comment:
I think we're doing more lifting than we need to do with the parsing we're doing here. Checking null, and looking for quotes is all work that the `json` package will do for us. We also don't need to parse out the full date with a regex manually given that the time library will do that for us. I think we should leverage `json.Unmarshal` if we can.

This is one approach:
```suggestion
	var s *string
	err := json.Unmarshal(data, &s)
	if err != nil {
		return err
	}
	if s == nil {
		return nil
	}
	parts := strings.SplitN(*s, "-", 2)
	year, err := strconv.ParseInt(parts[0], 10, 64)
	if err != nil {
		return err
	}
	rest, err := time.Parse(time.RFC3339, "0000-"+parts[1])
	if err != nil {
		return err
	}
	t.Time = rest.AddDate(int(year), 0, 0)
	return nil
```

Reply:
unfortunately splitting on the first "-" doesn't work on timestamps which have a negative year